crates/shim/src/cgroup.rs

```diff
@@ -199,6 +189,37 @@ fn get_cgroup(pid: u32) -> Result<Cgroup> {
     Ok(cgroup)
 }
 
+/// Get the cgroups v2 path given a PID
+pub fn get_cgroups_v2_path_by_pid(pid: u32) -> Result<String> {
+    // todo: should upstream to cgroups-rs
+    let path = format!("/proc/{}/cgroup", pid);
+    let content = fs::read_to_string(path).map_err(io_error!(e, "read cgroup"))?;
+    let content = content.trim_end_matches('\n');
+
+    parse_cgroups_v2_path(content)
+}
+
+// https://github.com/opencontainers/runc/blob/1950892f69597aa844cbf000fbdf77610dda3a44/libcontainer/cgroups/fs2/defaultpath.go#L83
+fn parse_cgroups_v2_path(content: &str) -> std::prelude::v1::Result<String, Error> {
+    // the entry for cgroup v2 is always in the format like `0::$PATH`
+    // where 0 is the hierarchy ID, the controller name is ommit in cgroup v2
+    // and $PATH is the cgroup path
+    // see https://docs.kernel.org/admin-guide/cgroup-v2.html
+    let parts: Vec<&str> = content.splitn(3, ":").collect();
+
+    if parts.len() < 3 {
+        return Err(Error::Other(format!("invalid cgroup path: {}", content)));
+    }
+
+    if parts[0] == "0" && parts[1].is_empty() {
+        // Check if parts[2] starts with '/', remove it if present.
+        let path = parts[2].strip_prefix('/').unwrap_or(parts[2]);
+        return Ok(format!("/sys/fs/cgroup/{}", path));
+    }
+
+    Err(Error::Other("cgroup path not found".into()))
```

Comment:
**suggestion:** I think this is slightly clearer:
```rust
let Some(path) = content.strip_prefix("0::") else {
    return Err(Error::Other(format!("invalid cgroup path: {}", content)));
}

let Ok(path) = PathBuf::from(format!("/sys/fs/cgroup/{path}")).canonicalize() {
    return Err(Error::Other("cgroup path not found".into()));
}

Ok(path)
```

Reply:
this actually makes it a bit hard to tests, for the time being, I will leave it as is for now. Is that fine?

Reply:
Good point.
How about this?
```rust
let Some(content) = content.lines().next() else {
    return Err(Error::Other("empty cgroup path".into()));
};

let Some(path) = content.strip_prefix("0::") else {
    return Err(Error::Other(format!("invalid cgroup path: {}", content)));
};

let path = path.trim_start_matches('/');

Ok(PathBuf::from(format!("/sys/fs/cgroup/{path}")))
```

Reply:
Done! Thanks